File: Namyata/OccurrencesAfterBigram.py
import unittest
from typing import List
# Given two strings first and second, consider occurrences in some text of the form "first second third",
# where second comes immediately after first, and third comes immediately after second.
#
# Return an array of all the words third for each occurrence of "first second third".



def findOcurrences(text: str, first: str, second: str) -> List[str]:
    words = text.split()
    thirds = []
    for word in range(len(words) - 2):
        if words[word] == first and words[word + 1] == second:
            thirds.append(words[word + 2])
    return thirds

class TestFindOcurrencesCorrected(unittest.TestCase):
    def test_empty_text(self):
        text = "There"
        first = "hello"
        second = "world"
        expected = []
        self.assertEqual(findOcurrences(text, first, second), expected)

    def test_no_occurrences(self):
        text = "This is a sample sentence."
        first = "hello"
        second = "world"
        expected = []
        self.assertEqual(findOcurrences(text, first, second), expected)

    def test_single_occurrence(self):
        text = "hello world hello there world"
        first = "hello"
        second = "world"
        expected = ["hello"]
        self.assertEqual(findOcurrences(text, first, second), expected)

    def test_multiple_occurrences(self):
        text = "hello world hello there world hello"
        first = "hello"
        second = "world"
        expected = ["hello"]
        self.assertEqual(findOcurrences(text, first, second), expected)

    def test_consecutive_occurrences(self):
        text = "hello world hello world hello"
        first = "hello"
        second = "world"
        expected = ["hello", "hello"]
        self.assertEqual(findOcurrences(text, first, second), expected)

    def test_long_text(self):
        text = "this is a long text with multiple occurrences of first and second words. first second first second first second."
        first = "first"
        second = "second"
        expected = ["first", "first"]
        self.assertEqual(findOcurrences(text, first, second), expected)


if __name__ == '__main__':
    unittest.main()


#Test cases ensured coverage but were wrong. Added the correct test cases.

File: CTests/occurrences_after_bigram.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// Function declaration
char** findOcurrences(const char* text, const char* first, const char* second, int* returnSize);

// Function implementation
char** findOcurrences(const char* text, const char* first, const char* second, int* returnSize) {
    const char* delimiter = " ";
    char* copy = strdup(text); // Make a copy of the text
    char* token = strtok(copy, delimiter);
    char** thirds = NULL;
    *returnSize = 0;

    while (token != NULL) {
        if (strcmp(token, first) == 0) {
            char* next = strtok(NULL, delimiter);
            if (next != NULL && strcmp(next, second) == 0) {
                char* third = strtok(NULL, delimiter);
                if (third != NULL) {
                    thirds = (char**)realloc(thirds, (*returnSize + 1) * sizeof(char*));
                    thirds[*returnSize] = strdup(third);
                    (*returnSize)++;
                }
            }
        }
        token = strtok(NULL, delimiter);
    }

    free(copy);
    return thirds;
}

// Test cases
void test_empty_text() {
    const char* text = "There";
    const char* first = "hello";
    const char* second = "world";
    int expectedSize = 0;
    char** expected = NULL;

    int returnSize;
    char** result = findOcurrences(text, first, second, &returnSize);

    printf("Test empty_text: ");
    if (returnSize == expectedSize) {
        printf("PASSED\n");
    } else {
        printf("FAILED\n");
    }

    // Free memory
    for (int i = 0; i < returnSize; i++) {
        free(result[i]);
    }
    free(result);
}

void test_no_occurrences() {
    const char* text = "This is a sample sentence.";
    const char* first = "hello";
    const char* second = "world";
    int expectedSize = 0;
    char** expected = NULL;

    int returnSize;
    char** result = findOcurrences(text, first, second, &returnSize);

    printf("Test no_occurrences: ");
    if (returnSize == expectedSize) {
        printf("PASSED\n");
    } else {
        printf("FAILED\n");
    }

    // Free memory
    for (int i = 0; i < returnSize; i++) {
        free(result[i]);
    }
    free(result);
}

void test_single_occurrence() {
    const char* text = "hello world hello there world";
    const char* first = "hello";
    const char* second = "world";
    int expectedSize = 1;
    char** expected = (char**)malloc(sizeof(char*) * expectedSize);
    expected[0] = strdup("hello");

    int returnSize;
    char** result = findOcurrences(text, first, second, &returnSize);

    printf("Test single_occurrence: ");
    if (returnSize == expectedSize) {
        int i;
        for (i = 0; i < returnSize; i++) {
            if (strcmp(result[i], expected[i]) != 0) {
                printf("FAILED\n");
                break;
            }
        }
        if (i == returnSize) {
            printf("PASSED\n");
        }
    } else {
        printf("FAILED\n");
    }

    // Free memory
    for (int i = 0; i < returnSize; i++) {
        free(result[i]);
    }
    free(result);
    for (int i = 0; i < expectedSize; i++) {
        free(expected[i]);
    }
    free(expected);
}

void test_multiple_occurrences() {
    const char* text = "hello world hello there world hello";
    const char* first = "hello";
    const char* second = "world";
    int expectedSize = 1;
    char** expected = (char**)malloc(sizeof(char*) * expectedSize);
    expected[0] = strdup("hello");

    int returnSize;
    char** result = findOcurrences(text, first, second, &returnSize);

    printf("Test multiple_occurrences: ");
    if (returnSize == expectedSize) {
        int i;
        for (i = 0; i < returnSize; i++) {
            if (strcmp(result[i], expected[i]) != 0) {
                printf("FAILED\n");
                break;
            }
        }
        if (i == returnSize) {
            printf("PASSED\n");
        }
    } else {
        printf("FAILED\n");
    }

    // Free memory
    for (int i = 0; i < returnSize; i++) {
        free(result[i]);
    }
    free(result);
    for (int i = 0; i < expectedSize; i++) {
        free(expected[i]);
    }
    free(expected);
}


int main() {
    test_consecutive_occurrences();
    test_multiple_occurrences();
    test_single_occurrence();
    test_no_occurrences();
    test_empty_text();
}
       
